### PR TITLE
Add e2e tests for Mounted s3 Drives

### PR DESCRIPTION
**Pull request by rlizzo**
## What does this PR do?

Adds an example and e2e test for mounting s3 drives in a `Work`. The test as written runs successfully against the platform staging env, but we will require a release before it can be run successfully in CI here. 

Fixes: GRID-9961

### Does your PR introduce any breaking changes? If yes, please list them.

No.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- ~[ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)~

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

_Yes, I had fun!_

cc @carmocca @akihironitta @borda @narJH27

### CODE REVIEW

**Comment by williamFalcon** [comment]
@rlizzo where are the docs for this feature? both the API docs and the workflow example

File: tests/tests_app_examples/test_drive_mounted_s3.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+from lightning_app.testing.testing import run_app_in_cloud
+
+
+@pytest.mark.cloud
+def test_drive_mounted_s3_example_cloud() -> None:
+    with run_app_in_cloud(os.path.join(_PROJECT_ROOT, "examples/app_drive_mounted_s3/"), app_name="app.py") as (
+        _,
+        _,
+        fetch_logs,
+        name,
```
Comment (Borda) [inline tests/tests_app_examples/test_drive_mounted_s3.py:17]:
```suggestion
        name
```

**Review by Borda — APPROVED** [review]

File: tests/tests_app_examples/test_drive_mounted_s3.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+from tests_app import _PROJECT_ROOT
+
+from lightning_app.cli.lightning_cli import logs
+from lightning_app.testing.testing import run_app_in_cloud
+
+
+@pytest.mark.cloud
+def test_drive_mounted_s3_example_cloud() -> None:
+    _path_s3 = os.path.join(_PROJECT_ROOT, "examples/app_drive_mounted_s3/")
+    with run_app_in_cloud(_path_s3), app_name="app.py") as (
```
Comment (Borda) [inline tests/tests_app_examples/test_drive_mounted_s3.py:14]:
```suggestion
    with run_app_in_cloud(_path_s3, app_name="app.py") as (
```

**Comment by rlizzo** [comment]
@williamFalcon docs have not been written yet. The plan is to validate the feature behaviors with Thomas when he returns and then roll out any changes he requests alongside the docs in the next `0.X` release of `lightning-app`

**Comment by manskx** [comment]
> The test as written runs successfully against the platform staging env, but we will require a release before it can be run successfully in CI here.

Why does it require a release? the CI runs against staging. @rlizzo

File: tests/tests_app_examples/test_drive_mounted_s3.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+@pytest.mark.cloud
+def test_drive_mounted_s3_example_cloud() -> None:
+    _path_s3 = os.path.join(_PROJECT_ROOT, "examples/app_drive_mounted_s3/")
+    with run_app_in_cloud(_path_s3, app_name="app.py") as (_, _, fetch_logs, name):
+        for _ in fetch_logs():
+            pass
+
+        runner = CliRunner()
+        result = runner.invoke(logs, [name])
+        lines = result.output.splitlines()
```
Comment (manskx) [inline tests/tests_app_examples/test_drive_mounted_s3.py:20]: Could reading the logs happen from the `fetch_logs()` ?

File: examples/app_drive_mounted_s3/app.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        # even though the main flow loop repeats infinitely if not explicitly exited.
+        self.work_1.run()
+        self.work_2.run(self.work2_drive_checkfile)
+
+        # are we done? If so, exit!
+        if self.work_1.done and self.work_2.done:
+            self._exit("Application End Successful!")
+
+        # If the works don't complete in < 30 minutes, a problem occured and we will exit here.
+        if time.time() - self.start_time > 30 * MINUTE:
```
Comment (manskx) [inline examples/app_drive_mounted_s3/app.py:158]: 30 minutes are too long to run a simple two works with the default image, I would say if they take more than 5 minutes this is an issue.

File: examples/app_drive_mounted_s3/app.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        os.makedirs(self.work2_drive_root_folder, exist_ok=True)
+
+        # create the drive and initialize Work2 with the drive object
+        drive = Drive("s3://ryft-public-sample-data/twitter/", root_folder=self.work2_drive_root_folder)
+        self.work_2 = Work2(drive, parallel=True)
+
+        # we will call Work2.run() with a file path to check to ensure that it is available and mounted.
+        self.work2_drive_checkfile = os.path.join(self.work2_drive_root_folder, "twitter2015.tar.gz")
+
+        self.start_time = None
```
Comment (manskx) [inline examples/app_drive_mounted_s3/app.py:141]: Could you shortly explain what this app does? 

I  would expect to see an example/test that the Drive is mounted to work_1 and writes some files, then work_2 to read the data of work_1.

File: examples/app_drive_mounted_s3/app.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        self.work_1.run()
+        self.work_2.run(self.work2_drive_checkfile)
+
+        # are we done? If so, exit!
+        if self.work_1.done and self.work_2.done:
+            self._exit("Application End Successful!")
+
+        # If the works don't complete in < 30 minutes, a problem occured and we will exit here.
+        if time.time() - self.start_time > 30 * MINUTE:
+            raise TimeoutError("More than 30 minutes elapsed since starting the flow")
```
Comment (awaelchli) [inline examples/app_drive_mounted_s3/app.py:159]: How long does it take to mount the S3 bucket though? Can it really be several minutes?

**Review by awaelchli — APPROVED** [review]

File: .github/workflows/ci-app-cloud-e2e-test.yml [hunk 7]
```diff
@@ -53,6 +53,7 @@ jobs:
           - collect_failures
           - custom_work_dependencies
           - drive
+          - drive_mounted_s3
```
Comment (hhsecond) [inline .github/workflows/ci-app-cloud-e2e-test.yml:56]: We are moving the e2e to [azure piplelines](https://github.com/Lightning-AI/lightning/pull/14282). You might have to add this to the new CI config file

Comment (hhsecond) [reply .github/workflows/ci-app-cloud-e2e-test.yml:56]: cc @Borda

Comment (Borda) [reply .github/workflows/ci-app-cloud-e2e-test.yml:56]: yep, let's see which PR gets merged first... :)

Comment (Borda) [reply .github/workflows/ci-app-cloud-e2e-test.yml]: updated

**Comment by Borda** [comment]
@rlizzo, mind checking the comments above? :rabbit:

File: examples/app_drive_mounted_s3/app.py (hunk 5 above)
Comment (tchaton) [reply examples/app_drive_mounted_s3/app.py:141]: This doesn't work with s3 drives. This is read only.

File: examples/app_drive_mounted_s3/app.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
+"""Demonstrates how a Drive can be used to mount an AWS s3 bucket to a Work's filesystem.
```
Comment (tchaton) [inline examples/app_drive_mounted_s3/app.py:1]: Could you move this example to the tests directly? There are too many asserts to be considered as an example.
Or would you mind merging this one with the other drive example?

File: examples/app_drive_mounted_s3/app.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        # Works which pass through the `parallel` argument to the L.Work base class run only once
+        # even though the main flow loop repeats infinitely if not explicitly exited.
+        self.work_1.run()
+        self.work_2.run(self.work2_drive_checkfile)
+
+        # are we done? If so, exit!
+        if self.work_1.done and self.work_2.done:
+            self._exit("Application End Successful!")
+
+        # If the works don't complete in < 30 minutes, a problem occured and we will exit here.
```
Comment (tchaton) [inline examples/app_drive_mounted_s3/app.py:157]: This isn't needed as the test would timeout.

**Comment by Borda** [comment]
@tchaton @rlizzo, what is the status here?

**Comment by Borda** [comment]
@tchaton @rlizzo, how is it going here? :rabbit:

**Comment by tchaton** [comment]
Hey @rlizzo, any updates? This isn't truly an example but a test. Let's move it under `tests/tests_app_examples/drive/`.

**Comment by Borda** [comment]
@tchaton @rlizzo, any update so far?

**Comment by Borda** [comment]
@narJH27 ^^

**Pull request closed without merging**